Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+            struct timespec ts;
+
+            _PyTime_t deadline = _PyTime_Add(_PyTime_GetSystemClock(), timeout);
+            _PyTime_AsTimespec(deadline, &ts);
+
+            err = pthread_cond_timedwait(&sema->cond, &sema->mutex, &ts);
+        }
+        else {
+            err = pthread_cond_wait(&sema->cond, &sema->mutex);
+        }
+        if (err) {
+            res = Py_PARK_TIMEOUT;
+        }
+    }
+    if (sema->counter > 0) {
+        sema->counter--;
+        res = Py_PARK_OK;
+    }
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+    return res;
+}
+
+int
+_PySemaphore_Wait(_PySemaphore *sema, _PyTime_t timeout, int detach)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    int was_attached = 0;
+    if (tstate) {
+        was_attached = (tstate->_status.active);
+        if (was_attached && detach) {
+            PyEval_ReleaseThread(tstate);
+        }
+    }
+
+    int res = _PySemaphore_PlatformWait(sema, timeout);
+
+    if (tstate) {
+        if (was_attached && detach) {
+            PyEval_AcquireThread(tstate);
+        }
+    }
+    return res;
+}
+
+void
+_PySemaphore_Wakeup(_PySemaphore *sema)
+{
+#if defined(MS_WINDOWS)
+    if (!ReleaseSemaphore(sema->platform_sem, 1, NULL)) {
+        Py_FatalError("parking_lot: ReleaseSemaphore failed");
+    }
+#elif defined(USE_SEMAPHORES)
+    int err = sem_post(&sema->platform_sem);
+    if (err != 0) {
+        Py_FatalError("parking_lot: sem_post failed");
+    }
+#else
+    pthread_mutex_lock(&sema->mutex);
+    sema->counter++;
+    pthread_cond_signal(&sema->cond);
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+}
+
+_PySemaphore *
+_PySemaphore_Alloc(void)
+{
+    // Make sure we have a valid thread_data. We need to acquire
+    // some locks before we have a fully initialized PyThreadState.
+    _PyParkingLot_InitThread();
+
+    ThreadData *this_thread = thread_data;
+    if (this_thread->depth >= MAX_SEMA_DEPTH) {
+        Py_FatalError("_PySemaphore_Alloc(): too many calls");
+    }
+    return &this_thread->semas[this_thread->depth++];
+}
+
+void
+_PySemaphore_Free(_PySemaphore *sema)
+{
+    ThreadData *this_thread = thread_data;
+    this_thread->depth--;
+    if (&this_thread->semas[this_thread->depth] != sema) {
+        Py_FatalError("_PySemaphore_Free(): mismatch wakeup");
+    }
+    _PyParkingLot_DeinitThread();
+}
+
+void
+_PyParkingLot_InitThread(void)
+{
+    if (thread_data != NULL) {
+        thread_data->refcount++;
+        return;
+    }
+    ThreadData *this_thread = PyMem_RawMalloc(sizeof(ThreadData));
+    if (this_thread == NULL) {
+        Py_FatalError("_PyParkingLot_InitThread: unable to allocate thread data");
+    }
+    memset(this_thread, 0, sizeof(*this_thread));
+    this_thread->refcount = 1;
+    this_thread->depth = 0;
+    for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
+        _PySemaphore_Init(&this_thread->semas[i]);
+    }
+    thread_data = this_thread;
+}
+
+void
+_PyParkingLot_DeinitThread(void)
```

Comment:
Likewise, consider "_PyParkingLot_ReleaseThread" [^1] or "_PyParkingLot_MaybeDeinitThread".

```suggestion
_PyParkingLot_ReleaseThread(void)
```

```suggestion
_PyParkingLot_MaybeDeinitThread(void)
```

[^1]: Perhaps that looks too similar to `PyEval_ReleaseThread()`?

Reply:
(removed)